packages/webpack-cli/lib/groups/ConfigGroup.js

```diff
@@ -58,162 +64,163 @@ const getConfigInfoFromFileName = (filename) => {
+        // when config is a function, pass the env from args to the config function
+        let formattedEnv;
+        if (Array.isArray(env)) {
+            formattedEnv = env.reduce((envObject, envOption) => {
+                envObject[envOption] = true;
+                return envObject;
+            }, {});
+        }
+        const newOptions = configOptions(formattedEnv, argv);
+        // When config function returns a promise, resolve it, if not it's resolved by default
+        newOptionsObject['options'] = await Promise.resolve(newOptions);
+    } else if (Array.isArray(configOptions) && configName) {
+        // In case of exporting multiple configurations, If you pass a name to --config-name flag,
+        // webpack will only build that specific configuration.
+        const namedOptions = configOptions.filter((opt) => configName.includes(opt.name));
+        if (namedOptions.length === 0) {
+            logger.error(`Configuration with name "${configName}" was not found.`);
+            process.exit(2);
+        } else {
+            newOptionsObject['options'] = namedOptions;
+        }
+    } else {
+        if (Array.isArray(configOptions) && !configOptions.length) {
+            newOptionsObject['options'] = {};
+            return newOptionsObject;
         }
+        newOptionsObject['options'] = configOptions;
     }
 
-    async resolveConfigMerging() {
-        const { merge } = this.args;
-        if (merge) {
-            // Get the current configuration options
-            const { options: configOptions } = this.opts;
-
-            // we can only merge when there are multiple configurations
-            // either by passing multiple configs by flags or passing a
-            // single config exporting an array
-            if (!Array.isArray(configOptions)) {
-                throw new ConfigError('Atleast two configurations are required for merge.', 'MergeError');
-            }
-
-            // We return a single config object which is passed to the compiler
-            const mergedOptions = configOptions.reduce((currentConfig, mergedConfig) => webpackMerge(currentConfig, mergedConfig), {});
-            this.opts['options'] = mergedOptions;
+    if (configOptions && configPath.includes('.webpack')) {
+        const currentPath = configPath;
+        const parentContext = dirname(currentPath).split(sep).slice(0, -1).join(sep);
+        if (Array.isArray(configOptions)) {
+            configOptions.forEach((config) => {
+                config.context = config.context || parentContext;
+            });
+        } else {
+            configOptions.context = configOptions.context || parentContext;
         }
+        newOptionsObject['options'] = configOptions;
     }
+    return newOptionsObject;
+};
 
-    async run() {
-        await this.resolveConfigFiles();
-        await this.resolveConfigMerging();
-        return this.opts;
+const resolveConfigMerging = async (args) => {
+    const { merge } = args;
```

Comment:
Isn't it better to use destructuring if the function takes in a single argument?

Reply:
neh I'll prefer this, looks cleaner